source/common/config/delta_subscription_state.cc

```diff
@@ -53,6 +55,31 @@ void DeltaSubscriptionState::updateSubscriptionInterest(
     names_added_.erase(r);
     names_removed_.insert(r);
   }
+  switch (mode_) {
+  case WildcardMode::Implicit:
+    if (names_removed_.find("*") != names_removed_.end()) {
+      // we explicitly cancel the wildcard subscription
+      mode_ = WildcardMode::Disabled;
+    } else if (!names_added_.empty()) {
```

Comment:
Hmm. I'm not sure about this: an explicit mode requires presence of a `*` in the names_added list; we only check if that list isn't empty. If another call to `updateSubscriptionInterest` is made before a request to the server is made, we may end up in a wrong state: for example if names_removed is empty in that call, we'll end up in `WildcardMode::Disabled` state.

Reply:
> Hmm. I'm not sure about this: an explicit mode requires presence of a `*` in the names_added list;

Only on the initial request (after the stream reset, for example), or on the first request where we decide to opt-in into the wildcard mode. When switching from implicit to explicit mode, no need to send `*` since the server should already know that we are in the wildcard mode. I would say that with this PR, the server should not ignore delta requests with a non-empty `resource_names_subscribe` list from a client that opted-in into wildcard mode.

I'll need to go through the documentation I wrote to make sure that it's clear.

> we only check if that list isn't empty. If another call to `updateSubscriptionInterest` is made before a request to the server is made, we may end up in a wrong state: for example if names_removed is empty in that call, we'll end up in `WildcardMode::Disabled` state.

Sorry, I'm not following here. There isn't any logic that resets us to `WildcardMode::Disabled` on an empty `names_removed_` list. Could you clarify?